app/Filament/Resources/AchievementResource/RelationManagers/EventAchievementsRelationManager.php

```diff
@@ -0,0 +1,135 @@
+                            })
+                            ->toArray();
+                    })
+                    ->getOptionLabelUsing(function (int $value): string {
+                        $achievement = Achievement::find($value);
+
+                        return "[{$achievement->id}] {$achievement->title}";
+                    }),
+
+                Forms\Components\DatePicker::make('active_from')
+                    ->label('Active From')
+                    ->native(false)
+                    ->date(),
+
+                Forms\Components\DatePicker::make('active_through')
+                    ->label('Active Through')
+                    ->native(false)
+                    ->date(),
+            ]);
+    }
+
+    public function table(Table $table): Table
+    {
+        /** @var User $user */
+        $user = Auth::user();
+
+        /** @var Achievement $achievement */
+        $achievement = $this->getOwnerRecord();
+
+        return $table
+            ->modifyQueryUsing(function (Builder $query) {
+                $query->with('sourceAchievement');
+            })
+            ->columns([
+                Tables\Columns\ImageColumn::make('sourceAchievement.badge_url')
+                    ->label('')
+                    ->size(config('media.icon.sm.width')),
+
+                Tables\Columns\TextColumn::make('sourceAchievement.title')
+                    ->label('Source Achievement')
+                    ->wrap(),
+
+                Tables\Columns\TextColumn::make('active_from')
+                    ->label('Active From')
+                    ->date(),
+
+                Tables\Columns\TextColumn::make('active_through')
+                    ->label('Active Through')
+                    ->date(),
+            ])
+            ->filters([
+
+            ])
+            ->headerActions([
+                Tables\Actions\CreateAction::make()
+                    ->createAnother(false)
+                    ->hidden($achievement->eventDatas()->count() > 0)
+                    ->after(function (EventAchievement $eventAchievement, Component $livewire) {
+                        $eventAchievement->refresh();
+                        (new CopyAchievementUnlocksToEventAchievement())->execute($eventAchievement);
+                        // TODO: refresh parent achievement data
+                    }),
+            ])
+            ->actions([
+                Tables\Actions\ActionGroup::make([
+                    Tables\Actions\EditAction::make()
+                        ->visible(function (EventAchievement $eventAchievement) use ($user) {
+                            return $user->can('update', $eventAchievement);
+                        })
+                        ->after(function (EventAchievement $eventAchievement, Component $livewire) {
+                            $eventAchievement->refresh();
+                            (new CopyAchievementUnlocksToEventAchievement())->execute($eventAchievement);
```

Comment:
It is my understanding we'll want this action to run any time `source_achievement_id`, `active_from`, and/or `active_until` on the event achievement are modified.

If that assumption is true, I think it will be much safer to use a Laravel model observer, which I've verified is working on my local:

```php
// app/Observers/EventAchievementObserver.php

<?php

declare(strict_types=1);

namespace App\Observers;

use App\Models\EventAchievement;
use App\Platform\Actions\CopyAchievementUnlocksToEventAchievement;

class EventAchievementObserver
{
    public function saved(EventAchievement $eventAchievement): void
    {
        $isFieldDirty = $eventAchievement->isDirty(['source_achievement_id', 'active_from', 'active_until']);
        $isConnectedToGameAchievement = $eventAchievement->achievement_id && $eventAchievement->source_achievement_id;

        if ($isFieldDirty && $isConnectedToGameAchievement) {
            (new CopyAchievementUnlocksToEventAchievement())->execute($eventAchievement);
        }
    }
}
```

```php
// app/Providers/EventServiceProvider.php

public function boot(): void
{
    // User::observe(UserObserver::class);
    EventAchievement::observe(EventAchievementObserver::class);
}
```

Now, Laravel will automatically execute the action any time one of those three fields changes. This is beneficial for two reasons:
1. Filament's sometimes-wacky API has areas where after-save hooks are not provided.
2. Regardless of how code gets changed or refactored in the future, we can trust that the observer will keep running.